File: Design Patterns/Structural/Decorator.ts
// lets you attach new behaviors to objects by placing these objects inside
// special wrapper objects that contain the behaviors


// ============================================================
// DECORATOR PATTERN — Coffee Shop
// Wraps an object to add behaviour without changing its class
// ============================================================

interface Coffee {
    cost(): number
    description(): string
}

// Base component
class SimpleCoffee implements Coffee {
    cost() { return 2 }
    description() { return "Coffee" }
}

// Decorators
abstract class CoffeeDecorator implements Coffee {
    constructor(protected coffee: Coffee) {}
    abstract cost(): number
    abstract description(): string
}

// Concrete decorators
class Milk extends CoffeeDecorator {
    cost() { return this.coffee.cost() + 0.5 }
    description() { return this.coffee.description() + ", Milk" }
}

class Sugar extends CoffeeDecorator {
    cost() { return this.coffee.cost() + 0.25 }
    description() { return this.coffee.description() + ", Sugar" }
}

class Vanilla extends CoffeeDecorator {
    cost() { return this.coffee.cost() + 0.75 }
    description() { return this.coffee.description() + ", Vanilla" }
}

let coffee: Coffee = new SimpleCoffee()
console.log(coffee.description(), `$${coffee.cost()}`)

coffee = new Milk(coffee)
console.log(coffee.description(), `$${coffee.cost()}`)

coffee = new Sugar(coffee)
coffee = new Vanilla(coffee)
console.log(coffee.description(), `$${coffee.cost()}`)

File: C#/Design Patterns/Structural/Decorator.cs
using System;

namespace Decorator
{
  interface ICoffee
  {
    decimal Cost();
    string Description();
  }

  class SimpleCoffee : ICoffee
  {
    public decimal Cost() { return 3.0m; }
    public string Description()
    { return "Ingredients: Coffee Beans"; }
  }

  abstract class CoffeeDecorator : ICoffee
  {
    protected ICoffee coffee;

    public CoffeeDecorator(ICoffee coffee)
    { this.coffee = coffee; }

    public abstract decimal Cost();
    public abstract string Description();
  }

  class Latte : CoffeeDecorator
  {
    public Latte(ICoffee coffee) : base(coffee) { }

    public override decimal Cost()
    { return coffee.Cost() + 0.75m; }

    public override string Description()
    { return $"{coffee.Description()}, Milk"; }
  }

  class Americano : CoffeeDecorator
  {
    public Americano(ICoffee coffee) : base(coffee) { }

    public override decimal Cost()
    { return coffee.Cost() + 0.50m; }

    public override string Description()
    { return $"{coffee.Description()}, Water"; }
  }

  class Program
  {
    public static void Main(string[] args)
    {
      ICoffee coffee = new SimpleCoffee();
      Console.WriteLine($"{coffee.Description()} | Cost: {coffee.Cost()}");

      coffee = new Latte(coffee);
      Console.WriteLine($"{coffee.Description()} | Cost: {coffee.Cost()}");

      coffee = new Americano(coffee);
      Console.WriteLine($"{coffee.Description()} | Cost: {coffee.Cost()}");
    }
  }

}